### PR TITLE
feat(GeosciencesIO): adding BRep fem writer

### PR DESCRIPTION
**Pull request by BenoitTHEBAULT**

### CODE REVIEW

File: src/geode/geosciences_io/mesh/fem_output.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        //     for( const auto v : values )
+        //     {
+        //         std::string line = "";
+        //         for( const auto e : att_dist[v] )
+        //         {
+        //             line += std::to_string( e + 1 ) + SPACE;
+        //         }
+        //         file_ << "     " << v << "  " << line << EOL;
+        //     }
+        // }
```
Comment (panquez) [inline src/geode/geosciences_io/mesh/fem_output.cpp:266]: to remove?

File: include/geode/geosciences_io/model/internal/brep_fem_output.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            }
+
+            static std::string_view extension()
+            {
+                static constexpr auto EXT = "fem";
+                return EXT;
+            }
+
+            std::vector< std::string > write( const BRep& brep ) const;
+            std::vector< std::string > write() const;
```
Comment (BotellaA) [inline include/geode/geosciences_io/model/internal/brep_fem_output.hpp:59]: How this one works?

File: src/geode/geosciences_io/mesh/fem_output.cpp [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+        void write_problem()
+        {
+            file_ << "PROBLEM:" << EOL;
+        }
+
+        void write_class()
+        {
+            file_ << "CLASS (v.7.202.18152)" << EOL;
+
```
Comment (BotellaA) [inline src/geode/geosciences_io/mesh/fem_output.cpp:86]: restrict the number of spaces

File: src/geode/geosciences_io/mesh/fem_output.cpp [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+            }
+        }
+
+        void write_xyz_dist()
+        {
+            write_one_coord_component_dist( "X", 0 );
+            write_one_coord_component_dist( "Y", 1 );
+            write_one_coord_component_dist( "Z", 2 );
+        }
+        // void write_node_property( geode::AttributeBase& attribute )
```
Comment (BotellaA) [inline src/geode/geosciences_io/mesh/fem_output.cpp:244]: to remove?

File: src/geode/geosciences_io/mesh/fem_output.cpp [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+
+            absl::flat_hash_map< double, std::vector< int > > att_dist =
+                create_att_dist( attribute );
+            std::vector< double > values;
+            values.reserve( att_dist.size() );
+            for( auto val = att_dist.begin(); val != att_dist.end(); ++val )
+            {
+                values.push_back( val->first );
+            }
+            std::sort( values.begin(), values.end() );
```
Comment (panquez) [inline src/geode/geosciences_io/mesh/fem_output.cpp:142]: is there an absl class to do it? (c_sort?)

File: src/geode/geosciences_io/mesh/fem_output.cpp [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+                  << " 0" << EOL;
+            for( const auto name :
+                solid_.polyhedron_attribute_manager().attribute_names() )
+            {
+                const auto attribute = solid_.polyhedron_attribute_manager()
+                                           .find_generic_attribute( name );
+                if( !attribute || !attribute->is_genericable()
+                    || name == "tetrahedron_vertices"
+                    || name == "tetrahedron_adjacents"
+                    || name == "geode_active" )
```
Comment (BotellaA) [inline src/geode/geosciences_io/mesh/fem_output.cpp:284]: why removing them?

**Review by BotellaA — APPROVED** [review]

File: src/geode/geosciences_io/mesh/fem_output.cpp [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+
+        void write_property( geode::AttributeBase& attribute )
+        {
+            file_ << attribute.name() << EOL;
+
+            absl::flat_hash_map< double, std::vector< int > > att_dist =
+                create_att_dist( attribute );
+            std::vector< double > values;
+            values.reserve( att_dist.size() );
+            for( auto val = att_dist.begin(); val != att_dist.end(); ++val )
```
Comment (panquez) [inline src/geode/geosciences_io/mesh/fem_output.cpp:138]: use python-like iteration?

File: src/geode/geosciences_io/model/internal/brep_fem_output.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE
+ * SOFTWARE.
```
Comment (panquez) [inline src/geode/geosciences_io/model/internal/brep_fem_output.cpp:26]: There are problems in header end of lines

File: src/geode/geosciences_io/model/internal/brep_fem_output.cpp [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                    .find_or_create_attribute< geode::VariableAttribute,
+                        std::string_view >( "Block_ID_polyhedron", "No name" );
+            auto attribute_v =
+                tet_solid.vertex_attribute_manager()
+                    .find_or_create_attribute< geode::VariableAttribute,
+                        std::string_view >( "Block_ID_vertex", "No name" );
+            for( const auto& block : brep_.blocks() )
+            {
+                for( const auto polyhedron_id :
+                    geode::Range( block.mesh().nb_polyhedra() ) )
```
Comment (panquez) [inline src/geode/geosciences_io/model/internal/brep_fem_output.cpp:94]: use { } after Ranges. Check everywhere

**Review by panquez — APPROVED** [review]

File: include/geode/geosciences_io/model/internal/brep_fem_output.hpp (hunk 2 above)
Comment (BenoitTHEBAULT) [reply include/geode/geosciences_io/model/internal/brep_fem_output.hpp:59]: En effet, c'était un artéfact!

File: src/geode/geosciences_io/mesh/fem_output.cpp (hunk 3 above)
Comment (BenoitTHEBAULT) [reply src/geode/geosciences_io/mesh/fem_output.cpp:86]: Dans les fichiers .fem, j'ai vui qu'il y en avait 4 ici, je peux en mettre moins?

File: src/geode/geosciences_io/mesh/fem_output.cpp (hunk 6 above)
Comment (BenoitTHEBAULT) [reply src/geode/geosciences_io/mesh/fem_output.cpp:284]: I did'nt find these attributes in fem files

**Pull request merged**